http/jsonapi/runtime/standart_params.go

```diff
@@ -0,0 +1,196 @@
+		return query
+	}, nil
+}
+
+// SortingFromRequest adds sorting to query based on the request query parameter
+// Database model and response type may differ, so the mapper allows to map the name of  field from the request
+// to a database column name
+// returns a (possible nop) queryOption, even if any sorting parameter was invalid
+// The error contains a list of all invalid parameters. Invalid parameters are not added to the query.
+func SortingFromRequest(r *http.Request, modelMapping map[string]string) (QueryOption, error) {
+	sort := r.URL.Query().Get("sort")
+	if sort == "" {
+		return func(query Query) Query { return query }, nil
+	}
+	sorting := strings.Split(sort, ",")
+
+	var order string
+	var resultedOrders []string
+	var errSortingWithReason []string
+	for _, val := range sorting {
+		if val == "" {
+			continue
+		}
+		order = " ASC"
+		if strings.HasPrefix(val, "-") {
+			order = " DESC"
+		}
+		val = strings.TrimPrefix(val, "-")
+
+		key, isValid := modelMapping[val]
+		if !isValid {
+			errSortingWithReason = append(errSortingWithReason, val)
+			continue
+		}
+		resultedOrders = append(resultedOrders, key+order)
+	}
+	sortingFilterOption := func(query Query) Query {
+		for _, val := range resultedOrders {
+			query.Order(val)
+		}
+		return query
+	}
+
+	if len(errSortingWithReason) != 0 {
+		return sortingFilterOption, fmt.Errorf("at least one sorting parameter is not valid: %q", strings.Join(errSortingWithReason, ","))
+	}
+	return sortingFilterOption, nil
+}
+
+// FilterFromRequest adds filter to a query based on the request query parameter
+// filter[name]=val1,val2 results in name IN (val1, val2), filter[name]=val results in name=val
+// Database model and response type may differ, so the mapper allows to map the name of field from the request
+// to a database column name and the sanitizer allows to correct type of the value and sanitize it.
+// Will always return a QueryOptions function with all valid filters (can be a nop)
+// if any filter are invalid a error with a list of all invalid filters is returned
+func FilterFromRequest(r *http.Request, modelMapping map[string]string, sanitizer ValueSanitizer) (QueryOption, error) {
+	filter := make(map[string][]interface{})
+	var invalidFilter []string
+	for queryName, queryValues := range r.URL.Query() {
+		if !(strings.HasPrefix(queryName, "filter[") && strings.HasSuffix(queryName, "]")) {
+			continue
+		}
+		key, isValid := getFilterKey(queryName, modelMapping)
+		if !isValid {
+			invalidFilter = append(invalidFilter, key)
+			continue
+		}
+		filterValues, isValid := getFilterValues(key, queryValues, sanitizer)
+		if !isValid {
+			invalidFilter = append(invalidFilter, key)
+			continue
+		}
+		filter[key] = filterValues
+	}
+
+	filterQueryOption := func(query Query) Query {
+		for name, filterValues := range filter {
+			if len(filterValues) == 0 {
+				continue
+			}
+
+			if len(filterValues) == 1 {
+				query.Where("?=?", name, filterValues[0])
```

Comment:
not sure it that works, could you add integration tests in general

Reply:
Still not Sure if this was okay, changed it before the test was ready